✨(vault) add a playbook to create a vault per application

lunika opened the pull request:
## Purpose

The vault files should be generated automatically when adding a customer

This PR closes #154 

## Proposal

Add a playbook to generate the vault files. We will use a convention to define where the template must be created (`apps/{{ app.name }}/vars/vault/main.yml.j2`).

- [x] add a playbook to generate vault files
- [x] create all default templates 
- [x] documentation

lunika committed:
✨(vault) add a playbook to create a vault per application

The playbook `create_vault.yml` will create all vaults configured in the
applications used by the client.

jmaupetit commented on create_vault.yml:
```diff
@@ -0,0 +1,18 @@
+
+  tasks:
+    - name: Display playbook name
+      debug: msg="==== Starting create_secrets playbook ===="
```
Copy/paste fail :wink:

jmaupetit commented on create_vault.yml:
```diff
@@ -0,0 +1,18 @@
+---
```
The file should be named `create_vaults.yml` because we will generate multiple vaults (one per app) for a project (customer/env_type).

jmaupetit commented on create_vault.yml:
```diff
@@ -0,0 +1,18 @@
+        tasks:
+          - tasks/create_app_vault
+      tags:
+        - vault
```
A carriage return is missing here.

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,31 @@
+    msg: "file apps/{{ app.name }}/vars/vault/main.yml.j2 does not exists"
+  when: vault_template.stat.exists == False
+
+
```
Extra empty row here.

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,31 @@
+- name: check if template file exists
+  fail:
+    msg: "file apps/{{ app.name }}/vars/vault/main.yml.j2 does not exists"
+  when: vault_template.stat.exists == False
```
:medal_sports:

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,31 @@
+  debug: msg="App name {{ app.name }}"
+  tags: vault
+
+- name: retrieve template file
```
Can you write all tasks name in sentence case (first word uppercase) please?

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,31 @@
+  when: vault_template.stat.exists == False
+
+
+- name: execute template to create vault file
```
Should be named: "Create application vault"

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,31 @@
+    - "apps/{{ app.name }}/vars/vault/main.yml.j2"
+
+- name: encrypt newly create vault file
+  local_action: command ansible-vault encrypt {{ item }}
```
I'd rather use the `command` module without local_action here:

```yaml
- name: Encrypt newly created vault file
  command: ansible-vault encrypt {{ item }}
  # ...
```

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,31 @@
+  local_action:
+    module: stat
+    path: "apps/{{ app.name }}/vars/vault/main.yml.j2"
+  register: vault_template
```
Use the `stat` module without enforcing the `local_action`.

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,31 @@
+  local_action:
+    module: template
+    src: "{{ item }}"
+    dest: "group_vars/customer/{{ customer }}/{{ env_type }}/secrets/{{ app.name }}.vault.yml"
```
Same here: remove the `local_action` call to use raw template module calls.

jmaupetit reviewed (commented):
I don't think we must enforce `local_action` usage, else :clap: for this first contribution :heart:

sampaccoud commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,36 @@
+    src: "{{ vault_template_path }}"
+    dest: "group_vars/customer/{{ customer }}/{{ env_type }}/secrets/{{ app.name }}.vault.yml"
+
+- name: Encrypt newly create vault file
```
s/create/created

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -5,27 +5,32 @@
+
+- name: Retrieve template file
+  stat:
+    path: "{{ vault_template_path }}"
```
Why not removing the two previous tasks and only keep this one as:

```yaml
- name: Get application vault template
  stat:
    path: "apps/{{ app.name }}/vars/vault/main.yml.j2"
  register: vault_template
```

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -5,27 +5,32 @@
+- name: Check if template file exists
   fail:
-    msg: "file apps/{{ app.name }}/vars/vault/main.yml.j2 does not exists"
+    msg: "file {{ vault_template_path }} does not exists"
```
This would then be: `vault_template.stat.path`

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -5,27 +5,32 @@
+- name: Encrypt newly create vault file
+  command: ansible-vault encrypt {{ item }}
   with_items:
     - "group_vars/customer/{{ customer }}/{{ env_type }}/secrets/{{ app.name }}.vault.yml"
```
As you are using it multiple times, define a variable for the application's vault path.

lunika committed:
✨(vault) use database vault to use credentials defined by user

In order to create vault by application, we use the databases
credentials present in a vault created before by the user

lunika committed:
🔧(vault) configure richie to generate its vault file

jmaupetit commented on apps/richie/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,12 @@
+# env_type: {{ env_type }}
+
+# postgresql
+{% set postgresql_credentials = postgresql | json_query("[?release=='" ~ richie_postgresql_version ~ "'].databases | [0][?application=='richie'].{user: user, password: password} | [0]")%}
```
A space is missing before the closing markup: `) %}`

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -5,6 +5,20 @@
   debug: msg="App name {{ app.name }}"
   tags: vault
 
+- name: Retrieve database vault file
```
`s/database/databases/`

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -5,6 +5,20 @@
+    msg: "database vault does not exists"
+  when: databases_vault.stat.exists == False
+
+- name: Import databases variable
```
> Import databases credentials

WDYT?

jmaupetit commented on apps/richie/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,12 @@
+{% set postgresql_credentials = postgresql | json_query("[?release=='" ~ richie_postgresql_version ~ "'].databases | [0][?application=='richie'].{user: user, password: password} | [0]")%}
+
+POSTGRESQL_USER: {{ postgresql_credentials.user }}
+POSTGRESQL_PASSWORD: {{ postgresql_credentials.password }}
```
The database name is missing.

jmaupetit replied:
My mistake: this var should not stand in the vault but in application vars.

sampaccoud commented on apps/richie/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,12 @@
+
+# richie
+DJANGO_SECRET_KEY: {{ lookup('password', '/dev/null length=30') }}
+# DJANGO_SENTRY_DSN: https://super:richie@sentry.io/foo
```
?

jmaupetit replied:
I think that this variable cannot be guessed, hence, for those kind of variables they can be left commented.

Maybe we must add a comment in the preamble or regroup them in a section with a comment explaining that they must be filled?

sampaccoud commented on tasks/create_app_vault.yml:
```diff
@@ -5,6 +5,20 @@
+
+- name: Check if databases vault exists
+  fail:
+    msg: "database vault does not exists"
```
s/exists/exist

lunika committed:
🔧(vault) configure edxapp to generate its vault file

lunika committed:
🔧(vault) configure forum to generate its vault file

lunika committed:
🔧(vault) configure marsha to generate its vault file

lunika committed:
📝(docs) complete playbook document with create_vaults.yml playbook

Explain how create_vaults.yml should be used and its purpose

lunika commented:
@jmaupetit @sampaccoud this PR is not in WIP anymore

jmaupetit commented on apps/edxapp/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,36 @@
+SECRETS_FILE:
+  CELERY_BROKER_USER: "{{ CELERY_BROKER_USER }}"
+  CELERY_BROKER_PASSWORD: "{{ CELERY_BROKER_PASSWORD }}"
+  COMMENTS_SERVICE_KEY: "thisisafakeapikey"
```
Add a comment for this variable, _e.g._:

```suggestion
  # FIXME: this COMMENTS_SERVICE_KEY should match the "forum" service API_KEY
  COMMENTS_SERVICE_KEY: "thisisafakeapikey"
```

jmaupetit commented on apps/edxapp/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,36 @@
+  MONGODB_NAME: "{{ MONGODB_DATABASE }}"
+  MONGODB_USER: "{{ MONGODB_USER }}"
+  MONGODB_PASSWORD: "{{ MONGODB_PASSWORD }}"
+{% endraw %}
```
You should configure your text editor to add a final newline. If you are using VSCode, add the following to your user settings:

```json
{
  "files.insertFinalNewline": true,
}
```

jmaupetit commented on apps/edxapp/vars/all/main.yml:
```diff
@@ -45,12 +45,14 @@ edxapp_mongodb_image_name: "centos/mongodb-32-centos7"
 edxapp_mysql_image_tag: "5.7"
 edxapp_mysql_port: 3306
 edxapp_mysql_secret_name: "edxapp-mysql-{{ edxapp_vault_checksum }}"
+edxapp_mysql_version: "5.7"
```
Same suggestion here:

```suggestion
edxapp_mysql_version: "5.7"
edxapp_mysql_image_name: "centos/mysql-57-centos7"
edxapp_mysql_image_tag: "{{ edxapp_mysql_version }}"
```

jmaupetit commented on apps/forum/vars/all/main.yml:
```diff
@@ -20,3 +20,4 @@ forum_mongodb_image_name: "centos/mongodb-32-centos7"
 forum_mongodb_image_tag: "3.2"
 forum_mongodb_port: 27017
 forum_mongodb_secret_name: "forum-mongo-{{ forum_vault_checksum }}"
+forum_mongodb_version: "3.2"
```
```suggestion
forum_mongodb_version: "3.2"
forum_mongodb_image_name: "centos/mongodb-32-centos7"
forum_mongodb_image_tag: "{{ forum_mongodb_version }}"
```

jmaupetit commented on apps/edxapp/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,36 @@
+MONGODB_ADMIN_PASSWORD: password
+MONGODB_USER: {{ mongodb_credentials.user }}
+MONGODB_PASSWORD: {{ mongodb_credentials.password }}
+MONGODB_DATABASE: {{ mongodb_credentials.name }}
```
Have you tried to make those variables `mandatory`?

```suggestion
MONGODB_DATABASE: {{ mongodb_credentials.name | mandatory }}
```

To ensure that if the corresponding variable is not defined in the databases vault, this templates fails to compile.

lunika replied:
I tried, it fails but the output error does not give information about which variable is mandatory. Not mentionning it as mandatory give the information about which variable is missing in the output error. I suggest to not use the mandatory filter

jmaupetit replied:
Ok then.

jmaupetit commented on apps/edxapp/vars/all/main.yml:
```diff
@@ -45,12 +45,14 @@ edxapp_mongodb_image_name: "centos/mongodb-32-centos7"
 edxapp_mongodb_image_tag: "3.2"
 edxapp_mongodb_port: 27017
 edxapp_mongodb_secret_name: "edxapp-mongo-{{ edxapp_vault_checksum }}"
+edxapp_mongodb_version: "3.2"
```
As the docker image tag matches required version, why not defining it only once, _e.g_:

```suggestion
edxapp_mongodb_version: "3.2"
edxapp_mongodb_image_name: "centos/mongodb-32-centos7"
edxapp_mongodb_image_tag: "{{ edxapp_mongodb_version }}"
```

lunika replied:
This doesn't work

jmaupetit commented on apps/forum/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,15 @@
+
+# Forum
+# Nota bene: the forum API_KEY should match the following edxapp lms settings:
+# COMMENTS_SERVICE_KEY
```
Add a "FIXME" comment here.

lunika replied:
I don't know what I can say more than the existing comment.

jmaupetit commented on apps/marsha/vars/all/main.yml:
```diff
@@ -16,6 +16,7 @@ marsha_postgresql_host: "postgresql"
 marsha_postgresql_port: 5432
 marsha_postgresql_database: "marsha"
 marsha_postgresql_secret_name: "marsha-postgresql-{{ marsha_vault_checksum }}"
+marsha_postgresql_version: "9.6"
```
```suggestion
marsha_postgresql_version: "9.6"
marsha_postgresql_image_name: "centos/postgresql-96-centos7"
marsha_postgresql_image_tag: "{{ marsha_postgresql_version }}"
```

lunika replied:
This doesn't work

jmaupetit commented on apps/marsha/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,23 @@
+
+# marsha environment
+DJANGO_SECRET_KEY: {{ lookup('password', '/dev/null length=30') }}
+# DJANGO_SENTRY_DSN: https://super:marsha@sentry.io/foo
```
Add a "FIXME" comment.

jmaupetit commented on apps/marsha/vars/vault/main.yml.j2:
```diff
@@ -0,0 +1,23 @@
+
+# AWS
+# DJANGO_AWS_ACCESS_KEY_ID: yourAwsAccessKeyId
+# DJANGO_AWS_SECRET_ACCESS_KEY: yourAwsSecretAccessKey
```
Add a "FIXME" comment for this section and the next one.

jmaupetit commented on apps/richie/vars/all/main.yml:
```diff
@@ -22,6 +22,7 @@ richie_postgresql_host: "postgresql"
 richie_postgresql_port: 5432
 richie_postgresql_database: "richie"
 richie_postgresql_secret_name: "richie-postgresql-{{ richie_vault_checksum }}"
+richie_postgresql_version: "9.6"
```
```suggestion
richie_postgresql_version: "9.6"
richie_postgresql_image_name: "centos/postgresql-96-centos7"
richie_postgresql_image_tag: "{{ richie_postgresql_version }}"
```

lunika replied:
This doesn't work

jmaupetit commented on docs/developer_guide/playbooks.md:
```diff
@@ -205,3 +205,18 @@ $ docker run --rm -it \
+application based on its template (`app/{app.name}/vars/vault/main.yml.j2`).
+This playbook will not work on all the applications present in your `apps` folder. Runs it only with a specific app.
+
+### usage
```
```suggestion
### Usage
```

jmaupetit commented on docs/developer_guide/playbooks.md:
```diff
@@ -205,3 +205,18 @@ $ docker run --rm -it \
+
+### usage
+
+you don't need to be connected to an openshift instance
```
```suggestion
You don't need to be connected to an OpenShift instance to run this playbook.
```

jmaupetit commented on docs/developer_guide/playbooks.md:
```diff
@@ -205,3 +205,18 @@ $ docker run --rm -it \
+This playbook will be used to create the vault files for a given application. It will reads the databases vault
+(must be present in `groups_vars/customers/{customer}/{env_type}/secrets/databases.ynl.vault`) and creates the vault
+application based on its template (`app/{app.name}/vars/vault/main.yml.j2`).
+This playbook will not work on all the applications present in your `apps` folder. Runs it only with a specific app.
```
Suggestion:

---

This playbook generates Ansible vaults (required to run applications) from templates bundled with the applications themselves (see `apps/{{ app.name }}/vars/vault/main.yml.j2`).

> Note that this playbook requires the `databases.vault.yml` vault as input to generate applications vaults. We expect this vault to be in `groups_vars/customers/{{ customer }}/{{ env_type }}/secrets/databases.vault.yml`.

jmaupetit commented on docs/developer_guide/playbooks.md:
```diff
@@ -205,3 +205,18 @@ $ docker run --rm -it \
+This playbook will be used to create the vault files for a given application. It will reads the databases vault
+(must be present in `groups_vars/customers/{customer}/{env_type}/secrets/databases.ynl.vault`) and creates the vault
+application based on its template (`app/{app.name}/vars/vault/main.yml.j2`).
+This playbook will not work on all the applications present in your `apps` folder. Runs it only with a specific app.
```
I don't understand your latest comment:

> This playbook will not work on all the applications present in your `apps` folder. Runs it only with a specific app. 

If you iterate over `apps` it should work out-of-box. Am I missing something?

lunika replied:
Yes it will not work because apps like `hello`, `mailcatcher` will not have the mandatory `vault/main.yml.j2` file

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,43 @@
+
+- name: Check if template file exists
+  fail:
+    msg: "file {{ vault_template.stat.path }} does not exists"
```
Can we be more explicit on this message? _e.g._:

> Required vault template for app {{ app.name }} does not exists.

Remember that `vault_template.stat.path` is not defined if `vault_template.stat.exists == False`

jmaupetit commented on tasks/create_app_vault.yml:
```diff
@@ -0,0 +1,43 @@
+- name: Create application vault
+  template:
+    src: "{{ vault_template.stat.path }}"
+    dest: "{{ vault_destination_path }}"
```
We should also prevent application's vault override:

```suggestion
- name: Get destination vault file status
  stat:
    path: "{{ vault_destination_path }}"
  register: vault_destination

- name: Create application vault
  template:
    src: "{{ vault_template.stat.path }}"
    dest: "{{ vault_destination_path }}"
  when: vault_destination.stat.exists == False
```

lunika commented:
once #163 done I will rework this one according new `databases.vault.yml` format

lunika commented:
@jmaupetit @sampaccoud This PR is rebased on master including #163 feature.

jmaupetit commented on docs/developer_guide/playbooks.md:
```diff
@@ -230,3 +230,18 @@ $ docker run --rm -it \
+
+## `create_vaults.yml`
+
+This playbook will be used to create the vault files needed by an application. It will reads the databases vault
```
```suggestion
This playbook will be used to create the vault files needed by an application. It will read the databases vault
```

jmaupetit reviewed (commented):
There is still unresolved comments to address :wink:

lunika commented:
:tada:
